### ISSUE TITLE
Ability to signal a running job that it's time to wrap-up and stop

### ISSUE THREAD

**Issue by golaz** [E3SM-Project/E3SM#4744]
I'm running a large ensemble on cori-knl by bundling multiple members into a single slurm allocation (job). It's not uncommon for one ensemble member to run substantially slower than others, causing many nodes to idle at the end of the allocation until the slow job finishes (or is killed).

Therefore, it would be really useful to have the ability to signal to a running instance of E3SM that it's time to stop. 

One possible mechanism to implement would be to add a special file in the run sub-directory. E3SM would periodically (every simulated day, or less frequently?) check that file and take action if needed. It would be useful to be able to tell a job to either stop ASAP or after the next scheduled set of restart files depending on user preference.

**Comment by rljacob** [comment]
So you want it to exit gracefully?   scancel JOBID can kill a slurm job immediately.   The check-a-file method is certainly easy to implement.   We would put it in the run directory for the instance you want to stop.

**Comment by golaz** [comment]
Yes, I'd like to exit more gracefully than `scancel`. That way, I would have restart files for the continuation job to restart from the previous one left off. Right now, I save restart files every years. If the job is killed in December, almost an entire year must be rerun. I could save restart files more frequently, but there is a cost to that as well (time and disk).

**Comment by philipwjones** [comment]
There are ways (in C,C++ and a Fortran wrapper to those) to catch some signals. You can't catch a kill signal (sort of by definition), but you can catch an abort or other signals. It could be hard to make it robust in a parallel application, especially if it happens to get caught in a region that's being synchronized, but maybe that can be managed in the signal handler. You'd certainly need to make sure a restart works correctly if called from a random time during a timestep update and make sure various time-averaging diagnostics are handled correctly. We had tried this in POP a long time ago, but found it was easier to use an odd/even restart strategy timed at a frequency typical of MTBF - and even that wasn't always perfect.

**Comment by golaz** [comment]
@philipwjones : we don't need to go as sophisticated as catching a Unix signal. I was suggesting something simpler: controlling via a text file within the `run/` sub-directory (e.g `STOP.txt`). Inside of that file, the user could specify specific instructions, for example:

- NEXT DAY/MONTH/YEAR : stop at 00 UTC of the next day, month or year
- NEXT RESTART : stop after writing the next scheduled restart

While running, e3sm.exe would periodically check that file and take action as needed.

**Comment by philipwjones** [comment]
Yeah, I was thinking more of the scancel approach above where you have an external monitor and external abort, so need to trap that for a more graceful exit. Handling it internally is certainly easier and your idea seems an easy way to implement. The one case this internal check doesn't catch is when there is a slowdown or hang due to a rank/node behaving badly and leaving the code in a hung/wait state.  In that case, the code itself isn't advancing and won't reach the internal check. For that we'd need an external kill signal from an outside monitor and a graceful exit would be impossible anyway. In any case, as we use more nodes for these ensembles the likelihood of encountering both hardware or simulation errors increases and we may need to think harder about building in fault tolerance.

**Comment by golaz** [comment]
I agree that an internal check won't take care of all cases, but it would be very useful nevertheless. Can someone implement this?

The fault tolerance piece is a different story. Right now, I'm trying to figure out how to recover manually by identifying and restarting failed ensemble members. If I find a reliable approach, I'll try to automate it.

**Comment by golaz** [comment]
@rljacob : can someone implement this?

**Comment by rljacob** [comment]
I was thinking of doing it because it would give me an excuse to look at the restart alarm logic.

**Comment by golaz** [comment]
@rljacob : are you still planning to work on this? Or would it be better to find someone with more time?

**Comment by rljacob** [comment]
We probably need to find someone else.

**Comment by golaz** [comment]
@singhbalwinder : do you think this is something you could do?

**Comment by singhbalwinder** [comment]
It has been a while since I looked at the alarm logic, but I can take a look at it and implement this logic.
